test_data/rules/unnecessary_stateful_widgets.dart

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2022, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `dart test -N unnecessary_stateful_widgets`
+
+import 'package:flutter/material.dart';
+
+class MyWidget extends StatefulWidget // LINT
+{
+  const MyWidget({super.key});
+
+  @override
+  State<MyWidget> createState() => _MyWidgetState();
+}
+
+class _MyWidgetState extends State<MyWidget> {
+  @override
+  Widget build(BuildContext context) {
+    return Container();
+  }
+}
+
+class GoodWidget extends StatelessWidget // OK
+{
+  const GoodWidget({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return Container();
+  }
+}
+
+// State with setState
+class MyWidget2 extends StatefulWidget // OK
+{
+  const MyWidget2({super.key});
+
+  @override
+  State<MyWidget2> createState() => _MyWidget2State();
+}
+
+class _MyWidget2State extends State<MyWidget2> {
+  @override
+  Widget build(BuildContext context) {
+    setState(() {});
+    return Container();
+  }
+}
+
+// State with field
+class MyWidget3 extends StatefulWidget // OK
+{
+  const MyWidget3({super.key});
+
+  @override
+  State<MyWidget3> createState() => _MyWidget3State();
+}
+
+class _MyWidget3State extends State<MyWidget3> {
```

Comment:
This should really be linted as well since the state is not really holding on to any state...

Reply:
But, of course, if a member `i` holds on to anything stateful (say a FocusNode or a ScrollController) it shouldn't lint. I guess, we probably can't differentiate between these cases?

Reply:
I started with only a rule to catch usage of `setState` but on the flutter codebase I face several false positives with State having fields. As you mentioned it's not easy to catch what field is stateful.

Reply:
If i remove the absence of fields in state condition, here's the list of diagnotics

```
   info • Unnecessary StatefulWidget • packages/flutter/lib/src/material/about.dart:530:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/lib/src/material/bottom_sheet.dart:67:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/lib/src/material/bottom_sheet.dart:350:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/lib/src/material/dropdown.dart:95:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/lib/src/widgets/fade_in_image.dart:61:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/test/material/reorderable_list_test.dart:1782:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/test/widgets/animated_list_test.dart:494:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/test/widgets/heroes_test.dart:162:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/test/widgets/inherited_model_test.dart:55:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/test/widgets/slivers_block_global_key_test.dart:11:7 • unnecessary_stateful_widgets
```

A lot of them are not simple/impossible to make statelessWidget